### PR TITLE
Feat(OptimisticConcurrency): WIP Adds expected version mechanism

### PR DESCRIPTION
**Pull request by pdt256**
Original PR #4

TODO:
- [x] Change X-ExpectedStreamSequenceNumber header?
- ~[ ] Add optimistic concurrency to Websocket API~ // WebSocket impl is for streaming only
- [x] upgrade to google.golang.org/protobuf
- [x] Move ExpectedStreamSequenceNumber to OptimisitcSave in remotestore and gRPC client/server
- ~[ ] Move helper string -> *uint64 to helper~ // Not necessary for private method
- [x] Deprecate SaveEvent
- [x] Add OptimisticSave
- [x] Update HTTP API docs to include ExpectedStreamSequenceNumber
- [x] Update gRPC docs to include ExpectedStreamSequenceNumber
- [x] Add tests for transaction rollback on SaveEvents failure
- [x] Add Optimistic Concurrency to all implementations
  - [x] LevelDB
  - [x] PostgreSQL
  - [x] Remote
  - [x] In-Memory

Optional:
- [x] Replace HTTP API save events to use proper json stream reading

### CODE REVIEW

**Comment by pdt256** [comment]
@Ganitzsh I had to revert some of your changes to get the tests to pass.

Thanks for finding the query issues in the PostgreSQL implementation. The SQL will need to be corrected again now that tests are passing. Also, the HTTP API needs to accommodate optimistic concurrency. Both of these issues will be resolved before merging this PR.

**Comment by pdt256** [comment]
More TODO:
- [ ] Change X-ExpectedStreamSequenceNumber header?
- [ ] Add optimistic concurrency to Websocket API
- [ ] upgrade to google.golang.org/protobuf
- [ ] Make ExpectedStreamSequenceNumber optional in protobuf
- [ ] Move helper string -> *uint64 to helper

**Review by yordis — APPROVED** [review]

File: pkg/rangedbapi/api.go [hunk 1]
```diff
@@ -165,7 +165,7 @@ func (a *api) saveEvents(w http.ResponseWriter, r *http.Request) {
 		return
 	}
 
-	expectedStreamSequenceNumber := stringToSequenceNumber(r.Header.Get("X-ExpectedStreamSequenceNumber"))
+	expectedStreamSequenceNumber := stringToSequenceNumber(r.Header.Get("ExpectedStreamSequenceNumber"))
```
Comment (yordis) [inline pkg/rangedbapi/api.go:168]: Isn't the usage of `X-` part of the standard for new headers?

Comment (pdt256) [reply pkg/rangedbapi/api.go:168]: Not anymore: https://tools.ietf.org/html/rfc6648

I thought about using `RDB-ExpectedStreamSequenceNumber`; however, I highly doubt there will ever be a collision with `ExpectedStreamSequenceNumber`.

Comment (yordis) [reply pkg/rangedbapi/api.go:168]: Wouldn't mind `RDB-*` thought, for a clear and easy way to map headers in proxies, potential tracing in a network and stuff like that.

Why would you do that? it is a matter of time.  🤷🏻

**Comment by pdt256** [comment]
Failing tests are due to this issue: https://github.com/codeclimate/test-reporter/issues/449

**Pull request merged**